Don't delete data, logs and config directories on uninstall

russcam opened the pull request:
- Delete install directory on uninstall and rollback, except data, config and logs directories when they are sub directories of the install directory.
- Split rollback and uninstallation directories tasks into two separate tasks, for clarity.

Closes #132

Mpdreamz commented on src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Rollback/RollbackDirectoriesTask.cs:
```diff
@@ -0,0 +1,132 @@
+
+			var fs = this.FileSystem;
+
+			if (this.Session.IsInstalling)
```
decrease nesting either by inverting the `if` or creating a method of the body

Mpdreamz replied:
Add a comment that this section will only be hit during a rollback of a fresh installation.

russcam replied:
> decrease nesting either by inverting the if or creating a method of the body

combined the `if`s

Mpdreamz commented on src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Uninstall/DeleteDirectoriesTask.cs:
```diff
@@ -0,0 +1,93 @@
+	/// We can't rely on the MSI RemoveFiles step to remove directories/files that were registered 
+	/// on a full uninstall because new files may have been introduced later on (e.g. plugins).
+	/// </summary>
+	public class DeleteDirectoriesTask : ElasticsearchInstallationTaskBase
```
`UninstallDeleteDirectoriesTask`

russcam replied:
I think `DeleteDirectoriesTask` is descriptive enough?

russcam replied:
renamed to `UninstallDirectoriesTask` to match the custom action

Mpdreamz commented on src/Tests/Elastic.Domain.Tests/Elasticsearch/Models/Tasks/Uninstall/UninstallDeleteDirectoriesTaskTests.cs:
```diff
@@ -23,11 +24,6 @@ public class UninstallDeleteDirectoriesTaskTests : InstallationModelTestBase
-				fs.Directory.Exists(m.LocationsModel.DataDirectory).Should()
-					.BeFalse("{0} {1}", m.LocationsModel.DataDirectory, session);
-				fs.Directory.Exists(m.LocationsModel.ConfigDirectory).Should().BeFalse("{0}", m.LocationsModel.ConfigDirectory);
-				fs.Directory.Exists(m.LocationsModel.LogsDirectory).Should().BeFalse("{0}", m.LocationsModel.LogsDirectory);
```
Should we not assert the opposite now rather than removing these assertations?

russcam replied:
Agreed 👍

Mpdreamz commented:
This PR for me still prevents the service to start without errors when going from `released 5.6.4` to this, the lingering plugins folder under `Elastic\Elasticsearch` is also still there after uninstall but that is unrelated to the two open PR's #142 and #141 

I think that we should validate on a `clean` install with no prior version that the selected data directory is empty/does not exist. If someone uninstalls `6.x` fully than tries to reinstall `5.x` they might get hard to debug situations. 

Should we rename the folders on uninstall instead?

russcam commented:
> This PR for me still prevents the service to start without errors when going from released 5.6.4 to this

@Mpdreamz Could you elaborate on "service to start without errors"? With this PR, the service starts successfully as part of an upgrade, but the subsequent uninstallation of 5.6.4 stops the newly installed service. There's no clean way of addressing this; the Restart Manager API _could_ be (ab)used from managed code to force a service restart after exiting the installer but I think the better approach is to document that the service will need to be started after upgrade. This issue should no longer occur with upgrades from this release onwards.

> the lingering plugins folder under Elastic\Elasticsearch is also still there after uninstall but that is unrelated 

will be addressed.

> I think that we should validate on a clean install with no prior version that the selected data directory is empty/does not exist. If someone uninstalls 6.x fully than tries to reinstall 5.x they might get hard to debug situations.

I think this is beyond the scope of the installer and something that would affect usage with any distribution; with such a check, the installer would not be able to discriminate between 

1. an older version that has been uninstalled (and the data, config, logs directories retained) and a newer version installed

    and

2. a newer version that has been uninstalled (and the data, config, logs directories retained) and an older version installed

russcam commented:
> > the lingering plugins folder under Elastic\Elasticsearch is also still there after uninstall but that is unrelated 
> 
> will be addressed.

On further thought, this shouldn't/can't be addressed; the uninstallation of 5.6.4 as part of an upgrade does not run the `DeleteDirectoriesTask`, on the assumption that the new version installed over the top of the existing version, including the plugins directory. With the change in #129 to include the version in the installation path and hence install to a different directory, this assumption no longer holds true.

russcam commented:
@Mpdreamz would you mind taking another look? I've made additional changes so that the DeleteDirectoriesTask runs on uninstall and is scheduled after the RemoveFolders action.

russcam commented:
rebased against master, following squash and merge of #141

russcam commented:
changed base of PR to target `master`

russcam commented:
I've rebased this PR against master following merge of #160, and squashed commits to make them more meaningful. I realise that some of the previous review comments may make little sense now.

@Mpdreamz would you be able to review? Specifically, the inclusion of Registry keys to persist the values of Config, Logs and Data directory paths. These need to be persisted so that the properties are populated with the correct values on uninstall.

russcam commented on src/Installer/Elastic.Installer.Msi/Elasticsearch/Elasticsearch.cs:
```diff
@@ -160,7 +188,7 @@ public override void PatchWixSource(XDocument document)
-				} 
+				}
 				// Add an empty plugins directory
 				else if (directoryId == "INSTALLDIR")
```
++

russcam commented on src/Installer/Elastic.Installer.Msi/Elasticsearch/Elasticsearch.cs:
```diff
@@ -82,10 +92,27 @@ public override void PatchProject(Project project)
+						$"[{msiParam.Attribute.Name}]")
+					{
+						AttributesDefinition = "Type=string",
+						Condition = $"(NOT {msiParam.Attribute.Name}=\"\")"
```
++ will remove

russcam commented on src/Installer/Elastic.Installer.Msi/Elasticsearch/Elasticsearch.cs:
```diff
@@ -19,11 +20,17 @@ public class Elasticsearch : Product
+		public override IEnumerable<ModelArgument> MsiParams => _msiParams ?? (_msiParams = 
+			ElasticsearchInstallationModel.Create(new NoopWixStateProvider(), NoopSession.Elasticsearch).ToMsiParams());
+
+		public override string RegistryKey => @"SOFTWARE\Elastic\Elasticsearch";
```
Went back and forward about using properties vs. product specific names. I think the property names would be cleaner. Will update.

russcam commented on src/Installer/Elastic.Installer.Msi/Program.cs:
```diff
@@ -27,12 +26,27 @@ private static void Main(string[] args)
+							RegistryHive.LocalMachine, 
+							product.RegistryKey, 
+							a.Key, 
+							a.Attribute.IsDynamic || a.Key == "INSTALLDIR" ? null : a.Value)
```
++ will define constant

codebrain commented on src/Installer/Elastic.Installer.Msi/Program.cs:
```diff
@@ -40,7 +41,7 @@ private static void Main(string[] args)
 							product.RegistryKey, 
 							a.Key, 
-							a.Attribute.IsDynamic || a.Key == "INSTALLDIR" ? null : a.Value)
+							a.Attribute.IsDynamic || a.Key == InstallDir ? null : a.Value)
```
:+1:

Mpdreamz commented on src/Installer/Elastic.Installer.Domain/Model/Kibana/Locations/LocationsModel.cs:
```diff
@@ -110,9 +110,9 @@ protected bool SamePathAs(string pathA, string pathB)
-				var fullPathA = Path.GetFullPath(pathA).TrimEnd('\\','/');
-				var fullPathB = Path.GetFullPath(pathB).TrimEnd('\\','/');
-				return 0 == string.Compare(fullPathA, fullPathB, true);
+				var fullPathA = Path.GetFullPath(pathA).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
```
👍

Mpdreamz commented on src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Uninstall/UninstallPluginsTask.cs:
```diff
@@ -1,48 +0,0 @@
-{
-	public class UninstallPluginsTask : ElasticsearchInstallationTaskBase
-	{
-		public UninstallPluginsTask(string[] args, ISession session) 
```
I know this works today because we have the `PreserveInstallAction` copying folders around but the fact we are sidestepping the `elasticsearch-plugin.bat` does not sit right with me. Although I realize this is now the way to go so no blocker.

We really need to make this an ES core concern e.g `bin/plugin reinstall` could simply update all official plugins currently installed or fail and rollback itself. then if we rollback as a whole on the installer we call `bin/plugin reinstall` itself again.

`</rant>`

russcam replied:
Agreed; it's a case of trying to fit an existing deployment model into the constraints of the MSI deployment model. The primary reasons that `elasticsearch-plugin.bat` can't be used are that 

1. the flow for updating plugins requires uninstall of current plugin versions and the install of new plugins
2. the plugins directory is not managed by the MSI, so files installed with plugin installation are not associated to components within the installer. Because there is no association, there is no file tracking in case of rollback.

Were `elasticsearch-plugin.bat` able to upgrade existing installed plugins, and rollback in case of error would be great and remove the need to copy the existing plugins directory on start of upgrade, to restore in case of rollback. Even if the responsibility were moved to `elasticsearch-plugin.bat`  however, I can't see how it would accomplish the rollback without caching the existing plugins somewhere, and a mechanism to purge this cache when necessary (since the upgrade process could fail at a later step, after successful plugin update).

An alternative approach would be bundling plugins with the installer as features, but this

1. would increase the size of the MSI database.
2. the database would need to have components for each plugin file and directory in order for tracking to work, which would require it to have knowledge of the deployment destination of each.
3. wouldn't help third party plugins, and what happens to those on upgrade.
